File: jbriareos-master/Tadeu_Freitas-jbriareos-c16b99c2c2d9/src/main/java/engine_modules/LeakDetector.java
package engine_modules;

import core.JMSInterface;
import engine_modules_helpers.IOData;
import engine_modules_helpers.IOType;
import engine_modules_helpers.ModuleIO;
import engine_modules_helpers.ModuleType;
import org.apache.commons.lang.ArrayUtils;
import utils.BPacket;
import utils.Common;

import java.io.BufferedReader;
import java.io.FileReader;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.*;
import java.util.stream.Stream;

public class LeakDetector extends EngineModule {
    public LeakDetector() {
        super(ModuleType.LeakDetector);
    }

    @Override
    public IOType getInputType() {
        return IOType.Tuple;
    }

    @Override
    public IOType getOutputType() {
        return IOType.Tuple;
    }

    @Override
    public Common.InputMode getInputMode() {
        return Common.InputMode.SingleInput;
    }

    private static final int MIN_LEAK_LENGTH = 2;
    private static final int REPLACE = 0;
    private static final int BLOCK = 1;
    private static final int ALLOW_EXPLOIT = 2;
    private static final int MODE = BLOCK;


    private static class MemMap {
        int[] addressRange;
        String pattern;
        String permissions;
        String name;
        String[] addressRangeHex;

        public MemMap(int[] addressRange, String pattern, String permissions, String name, String[] addressRangeHex) {
            this.addressRange = addressRange;
            this.pattern = pattern;
            this.permissions = permissions;
            this.name = name;
            this.addressRangeHex = addressRangeHex;
        }
    }

    private String getPattern(String a, String b) {
        StringBuilder pattern = new StringBuilder();
        a = a.replace("\\x00", "");
        b = b.replace("\\x00", "");

        int minLength = Math.min(a.length(), b.length());
        for (int i  = 0; i < minLength; i++) {
            if (a.charAt(i) == b.charAt(i))
                pattern.append(a.charAt(i));
            else break;
        }

        return pattern.toString();
    }

    private LinkedList<MemMap> getMemoryMaps(long pid) {
        LinkedList<MemMap> maps = new LinkedList<>();

        try {
            BufferedReader reader = new BufferedReader(new FileReader("/proc/" + pid + "/maps"));
            String line = reader.readLine();
            while (line != null) {
                line = line.trim();
                String[] l = line.split(" ");

                String[] lAddr = l[0].split("-");
                int[] addressRange = {Integer.parseInt(lAddr[0],16), Integer.parseInt(lAddr[1],16)};
                String[] addressRangeHex = {Integer.toHexString(addressRange[0]), Integer.toHexString(addressRange[1])};

                ByteBuffer buf = ByteBuffer.allocate(4);
                buf.order(ByteOrder.LITTLE_ENDIAN);
                buf.putInt(addressRange[0]);
                byte[] patStartAddr = buf.array();
                ArrayUtils.reverse(patStartAddr);

                ByteBuffer buf1 = ByteBuffer.allocate(4);
                buf1.order(ByteOrder.LITTLE_ENDIAN);
                buf1.putInt(addressRange[1]);
                byte[] patEndAddr = buf1.array();
                ArrayUtils.reverse(patEndAddr);

                String pattern = getPattern(Arrays.toString(patStartAddr), Arrays.toString(patEndAddr));
                if (pattern.length() >= MIN_LEAK_LENGTH) {
                    String permissions = l[1];
                    String name = "Unknown";
                    if (l.length == 6)
                        name = l[5];

                    maps.addLast(new MemMap(addressRange, pattern, permissions, name, addressRangeHex));
                }

                line = reader.readLine();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }

        return maps;
    }

    @Override
    public ModuleIO process(BPacket packet, ModuleIO data, JMSInterface jmsInterface) {
        if (data.ioType != IOType.Tuple) {
            System.out.println("-- LeakDetector --");
            System.out.println("Data's type is not the same as the data type needed by the module");
            System.exit(-1);
        }

        long pid = ((IOData.IOTuple) data.ioData).number;

        String packetData = packet.getApplicationData();
        if (packetData.isEmpty()) {
            packet.accept();
            return data;
        }

        Optional<ProcessHandle> obj = ProcessHandle.of(pid);
        if (!obj.isPresent()) {
            packet.accept();
            return data;
        }

        ProcessHandle process = obj.get();
        Stream<ProcessHandle> children = process.children();

        Stream<ProcessHandle> processStream = Stream.of(process);
        Iterator<ProcessHandle> iterator = Stream.concat(processStream, children).iterator();
        while (iterator.hasNext()) {
            ProcessHandle ph = iterator.next();
            LinkedList<MemMap> maps = getMemoryMaps(ph.pid());

            char[] packetDataArray = packetData.toCharArray();
            ArrayUtils.reverse(packetDataArray);

            String reversePacketData = Arrays.toString(packetDataArray);
            for (MemMap memMap : maps) {
                if (reversePacketData.contains(memMap.pattern)) {
                    if (MODE == BLOCK) {
                        System.out.println("Leak detected " + memMap.name +
                                " (" + Arrays.toString(memMap.addressRangeHex) + ") " +
                                "-> Dropping packet");
                        packet.drop();
                        return data;
                    } else if (MODE == REPLACE) {
                        System.out.println("Leak detected " + memMap.name + " " +
                                "-> Replacing leak data");
                        return data;
                    } else {
                        System.out.println("Leak detected " + memMap.name + " " +
                                "-> Waiting for exploit to complete");
                        break;
                    }
                }
            }
        }

        return data;
    }
}

File: briareos-master/dev/project/engine/modules/leak_detector.py
# Pipeline Module

from briareos.utils.modules import *
from briareos.utils import logger

import psutil

# Options: block and log, replace and log, allow exploit and save exploit?

REPLACE = 0
BLOCK = 1
ALLOW_EXPLOIT = 2

MODE = BLOCK
INCLUDE_FORKS = True
STRUCT_FORMAT = "<Q"  # 64 bits little-endian
MIN_LEAK_LENGTH = 2


class LeakDetector(Module):
    name = "Address Leak Detector"
    description = "Prevents 0day exploits by blocking leaks of memory addresses"
    input_type = tuple
    output_type = tuple

    # TODO replace - API
    # TODO Extra mode, let the exploit run but save it for fun and profit! -> relacionar Input/Output
    # TODO get canary leaks https://www.elttam.com.au/blog/playing-with-canaries/
    # TODO more efficient
    # TODO check if exploit is complete
    
    def process(self, packet, process_info):
        pid, app_name = process_info

        data = packet.get_application_data()
        if data is None:
            packet.accept()
            return process_info

        process = psutil.Process(pid=pid)
        children = process.children(recursive=True)

        for proc in [process] + children:
            maps = get_memory_maps(proc.pid)
            rev_data = data[::-1]

            for memory_map in maps:
                pattern = memory_map[1]
                if pattern in rev_data:
                    if MODE == BLOCK:
                        logger.warning("Leak detected %s (%s) -> Dropping packet" % (memory_map[3], memory_map[4]))
                        # exploitDetected()
                        packet.drop(final=True)
                        return process_info
                    elif MODE == REPLACE:
                        logger.warning("Leak detected %s -> Replacing leak data" % memory_map[3])
                        return process_info
                    elif MODE == ALLOW_EXPLOIT:
                        logger.warning("Leak detected %s -> Waiting for exploit to complete" % memory_map[3])
                        # exploitDetected()
                        break

        return process_info


def get_pattern(a, b):
    pattern = ""
    a = a.replace("\x00", "")
    b = b.replace("\x00", "")
    for i in range(min(len(a), len(b))):
        if a[i] == b[i]:
            pattern += a[i]
        else:
            break
    return pattern


def get_memory_maps(pid):
    maps = []

    with open("/proc/%s/maps" % pid) as f:
        lines = f.readlines()
        for line in lines:
            line = line.strip()
            l = line.split()

            l_addr = l[0].split("-")
            address_range = (int(l_addr[0], 16), int(l_addr[1], 16))
            address_range_hex = (hex(int(l_addr[0], 16)), hex(int(l_addr[1], 16)))
            pat_start_addr = struct.pack(STRUCT_FORMAT, address_range[0])[::-1]
            pat_end_addr = struct.pack(STRUCT_FORMAT, address_range[1])[::-1]

            pattern = get_pattern(pat_start_addr, pat_end_addr)

            if len(pattern) >= MIN_LEAK_LENGTH:
                permissions = l[1]
                if len(l) == 6:
                    name = l[5]
                else:
                    name = "Unknown"
                maps.append((address_range, pattern, permissions, name, address_range_hex))

    return maps
